app/src/main/java/com/kpiroom/bubble/ui/accountSetup/AccountSetupLogic.kt

```diff
@@ -0,0 +1,182 @@
+package com.kpiroom.bubble.ui.accountSetup
+
+import android.net.Uri
+import androidx.lifecycle.MutableLiveData
+import com.kpiroom.bubble.R
+import com.kpiroom.bubble.os.BubbleApp.Companion.app
+import com.kpiroom.bubble.source.Source
+import com.kpiroom.bubble.source.api.impl.firebase.FirebaseStructure.User
+import com.kpiroom.bubble.ui.core.CoreLogic
+import com.kpiroom.bubble.util.async.AsyncProcessor
+import com.kpiroom.bubble.util.constants.getResUri
+import com.kpiroom.bubble.util.constants.str
+import com.kpiroom.bubble.util.databinding.ProgressState
+import com.kpiroom.bubble.util.livedata.*
+import kotlinx.coroutines.Dispatchers
+import java.text.SimpleDateFormat
+import java.util.*
+
+class AccountSetupLogic : CoreLogic() {
+    val progress = MutableLiveData<ProgressState>()
+
+    val username = MutableLiveData<String>()
+    private val usernameRegex = str(R.string.username_regex).toRegex()
+    val clearUsernameFocus = MutableLiveData<Boolean>()
+
+    private val dateFormat = str(R.string.date_format)
+    private val currentDateString = SimpleDateFormat(dateFormat).format(Date()).toString()
+    val joinedOn = "${str(R.string.setup_joined_on)} $currentDateString"
+
+    val photoChangeRequested = MutableLiveData<Boolean>()
+    val wallpaperChangeRequested = MutableLiveData<Boolean>()
+
+    val finishRequested = MutableLiveData<Boolean>()
+    val setupComplete = MutableLiveData<Boolean>()
+
+    private val defaultPhotoUri = getResUri(R.drawable.default_avatar)
+    val photoUri = MutableLiveData<Uri>().setDefault(defaultPhotoUri)
+
+    private val defaultWallpaperUri = getResUri(R.drawable.default_wallpaper)
+    val wallpaperUri = MutableLiveData<Uri>().setDefault(defaultWallpaperUri)
+
+    private val isPhotoSet
+        get() = photoUri.value != defaultPhotoUri
+    private val isWallpaperSet
+        get() = wallpaperUri.value != defaultWallpaperUri
+
+    val backButtonClicked = MutableLiveData<Boolean>()
+    fun onBackButton() {
+        Source.userPrefs.clear()
+        backButtonClicked.value = true
+    }
+
+    fun onPhotoChanged() {
+        clearUsernameFocus.value = true
+        photoChangeRequested.value = true
+        wallpaperChangeRequested.value = false
+    }
+
+    fun onWallpaperChanged() {
+        clearUsernameFocus.value = true
+        photoChangeRequested.value = false
+        wallpaperChangeRequested.value = true
+    }
+
+    fun onGetStarted() {
+        clearUsernameFocus.value = true
+        progress.apply {
+            username.value.let {
+
+                AsyncProcessor {
+                    progress.loadAsync()
+                    when {
+                        it.isNullOrBlank() -> alertAsync(str(R.string.setup_username_cannot_be_empty))
+
```

Comment:
Split this line.
More details: https://kotlinlang.org/docs/reference/coding-conventions.html in `where` statement

Reply:
What is a `where` statement?

Reply:
*`whene`